src/main/java/com/firebase/tubesock/WebSocket.java

```diff
@@ -344,9 +292,97 @@ private void verifyHost(SSLSocket socket, String host) throws SSLException {
      */
     public void blockClose() throws InterruptedException {
         // If the thread is new, it will never run, since we closed the connection before we actually connected
-        if (writer.getState() != Thread.State.NEW) {
-            writer.join();
+        if (writer.getInnerThread().getState() != Thread.State.NEW) {

```

Comment:
there is a subtle race condition here, that if you call blockClose() immediately after start() you might call getInnerThread() on two different threads resulting in two different Threads created. I would propose the following changes:
a) Make a runWriter and runReader method respectively
b) Create the threads in the constructors and assign them to a local final variable
c) Make getInnerThread package private
d) Return a local final variable in the getInnerThread